src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -1217,94 +1140,107 @@ protected void shutdownService() {
-                if (scheduleQueue.remove(nextScheduledTask)) {
-                  nextScheduledTask.executing();
-                  return nextScheduledTask;
-                }
+        TaskWrapper nextHighPriorityTask = highPriorityQueueSet.getNextTask();
+        TaskWrapper nextLowPriorityTask = lowPriorityQueueSet.getNextTask();
+        if (nextLowPriorityTask == null) {
+          if (nextHighPriorityTask == null) {
+            // no tasks, so just wait till any tasks come in
+            LockSupport.park();
+          } else {
+            // only high priority task is queued, so run if ready
+            long nextTaskDelay = nextHighPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+            if (nextTaskDelay <= 0) {
+              if (nextHighPriorityTask.canExecute()) {
+                return nextHighPriorityTask;
               }
-            } else if (executeQueue.remove(nextExecuteTask)) {
-              // if we can remove the task (aka it has not been removed already), we can execute it
-              nextExecuteTask.executing();
-              return nextExecuteTask;
+            } else {
+              LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextTaskDelay);
             }
-          } else if (executeQueue.remove(nextExecuteTask)) {
-            // if we can remove the task (aka it has not been removed already), we can execute it
-            nextExecuteTask.executing();
-            return nextExecuteTask;
           }
-        } else if (nextScheduledTask != null) {
-          if (nextScheduledTask.getDelay(TimeUnit.MILLISECONDS) <= 0) {
-            synchronized (scheduleQueue.getModificationLock()) {
-              // scheduled tasks must be removed, and call .executing() while holding the lock
-              if (scheduleQueue.remove(nextScheduledTask)) {
-                nextScheduledTask.executing();
-                return nextScheduledTask;
-              }
+        } else if (nextHighPriorityTask == null) {
+          // only low priority task is queued, so run if ready
+          long nextTaskDelay = nextLowPriorityTask.getDelay(TimeUnit.MILLISECONDS);
+          if (nextTaskDelay <= 0) {
+            if (nextLowPriorityTask.canExecute()) {
+              return nextLowPriorityTask;
             }
           } else {
-            workerPool.handleEstimatedTimeTillNextTask(this, nextScheduledTask.getDelayEstimateInMs());
-            LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextScheduledTask.getDelay(TimeUnit.MILLISECONDS));
+            LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextTaskDelay);
           }
         } else {
-          workerPool.handleEstimatedTimeTillNextTask(this, Long.MAX_VALUE);
-          LockSupport.park();
+          // both tasks are available, so see which one makes sense to run
+          long nextHighDelay;
+          long nextLowDelay;
+          ClockWrapper.stopForcingUpdate();
```

Comment:
We use the wrapper not for performance, but to prevent unfairly choosing (or inserting) one task incorrectly due to clock changes during inspection.  This is actually a deficency in the java.util.concurrent scheduler that we have solved.

I will play with the ordering of these a bit though.